old-tutorials-in-development/tutorials-in-development/R/dc-spatio-temporal-intro/09-tabular-data-bad-missing-values.Rmd

```diff
@@ -0,0 +1,361 @@
+with this dataset for more information on this quality flag column. 
+
+``` {r  view-metadata, eval = FALSE}
+
+# read metadata csv
+vegMetadata <- read.csv(paste0(wd,"NEON-DS-Field-Site-Spatial-Data/SJER/VegetationData/D17_2013_vegStr_metadata_desc.csv"))
+
+# view the dataframe
+View(vegMetadata)
+
+```
+
+Now that we know there are bad values in our data, we can start to deal with 
+them appropriately. 
+
+### Deal with Bad Data Values
+As with handling missing values, there are different ways to deal with bad data 
+values. One way is to change those values into `NA` or some other non-integer 
+value so `R` does not recognize it in mathematical operations.
+
+Before changing the zeroes in the `dbhheight` column, let's make sure there aren't 
+actual `NA` values. If there are no `NA` values to take into consideration, then 
+we can change the bad values into `NA` to make it easier to remove them from 
+the dataframe.
+
+``` {r check-and-replace-NA}
+
+# check for NA values in dbhheight
+sum(is.na(vegData$dbhheight))
+
+# change zeroes in dbhheight to NA
+vegData$dbhheight[vegData$dbhheight==0] <- NA
+
+```
+
+Using the `replace_na` function with pipes (`%>%`) from the `tidyr` package, we 
+can replace the `NA` values we just created with `ERROR` to make it obvious that 
+these are bad data values.  
+
+``` {r NA-to-error}
+
+# load library
+library(tidyr)
+
+# replace the NAs with "error"
+vegData$dbhheight %>% replace_na("ERROR")
+
+# if desired, create a new dataframe with the "error" values
+vegErrors <- vegData %>% replace_na(list(dbhheight="ERROR"))
+
+```
+
+<i class="fa fa-star"></i> **Data Tip:** The `tidyr` package won't modify the 
+data in the dataframe and the `replace_na` function will only show a view of the 
+modified data. In order to make those changes permanent, create a new dataframe 
+or assign the modifications to the original data: 
+`vegData <- vegData %>% replace_na("ERROR")`
+
+When dealing with `NoData` values in a previous section, we used `na.omit` to 
+create a new dataframe without **any** `NA` values. Below, use the `complete.cases` 
+function to remove rows containing `NA` values **only** from specified columns. 
+Alternatively, use another function from the `tidyr` package to remove `NA` values. 
+
+``` {r remove-dbhheight-NAs, eval = FALSE}
+
+# remove NAs only from dbhheight column
+vegData2 <- vegData[complete.cases(vegData[,13]),] #dbhheight is column 13
+
+# or use drop_na from the tidyr package
+drop_na(vegData, "dbhheight")
+
+```
+
+Packages such as `tidyr` and `dplyr` are powerful tools for data cleaning and 
+manipulation. This article on <a href="https://towardsdatascience.com/data-cleaning-with-r-and-the-tidyverse-detecting-missing-values-ea23c519bc62" target="_blank">Data Cleaning with R and the Tidyverse </a>
+goes into more detail on these packages.
+
+In addition to finding `NoData` or bad data values, sometimes we might need to 
```

Comment:
@donal-at-NEON I moved this section (351-356) from where the challenge section is above to down here because I think it makes more sense at the end.

Reply:
I agree, this is a good move. Thanks!